### PR TITLE
[FSDP1] fix _same_storage check for DTensor

### PR DESCRIPTION
**Pull request by weifengpy**
for FSDP (SHARD_GRAD_OP + use_orig_params) + TP, params in the backward are DTensors. However,  ``DTensor.untyped_storage().data_ptr()`` does not work in ``_same_storage``. Thus desugar to ``DTensor._local_tensor.untyped_storage().data_ptr()`` https://github.com/pytorch/pytorch/issues/123272

credit to @bigning for the original fix. after landing, we would not need patching in mosaic composer https://github.com/mosaicml/composer/pull/3175/files

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang

### CODE REVIEW

File: torch/distributed/fsdp/_flat_param.py [hunk 1]
```diff
@@ -27,6 +27,7 @@
 import torch.nn as nn
 import torch.nn.functional as F
 from torch import Tensor
+from torch.distributed._tensor import DTensor
```
Comment (awgu) [inline torch/distributed/fsdp/_flat_param.py:30]: This is not a great state to be in, but I always remember that we cannot import `DTensor` at the top-level of this file, or else we may break some internal torch package or torch deploy thing.

I am not too familiar with the issue though :/

File: test/distributed/fsdp/test_fsdp_tp_integration.py [hunk 2]
```diff
@@ -141,31 +141,36 @@ def _sync_tp_grads(
         tp_world_size = tp_pg.size()
         fsdp_world_size = self.world_size // tp_world_size
         assert (
-            type(tp_fsdp_model) is FSDP and len(list(tp_fsdp_model.parameters())) == 1
+            type(tp_fsdp_model) is FSDP
```
Comment (awgu) [inline test/distributed/fsdp/test_fsdp_tp_integration.py:144]: IIUC, this change is to make the check stricter to more accurately reflect our assumptions?

File: test/distributed/fsdp/test_fsdp_tp_integration.py [hunk 3]
```diff
@@ -182,7 +187,14 @@ def _get_grads_as_flattened(
         returns the same value on all ranks.
         """
         local_grads_as_flattened = (
-            torch.cat([torch.flatten(param.grad) for param in model.parameters()])
+            torch.cat(
+                [
+                    torch.flatten(param.grad)
+                    if param.grad is not None
```
Comment (awgu) [inline test/distributed/fsdp/test_fsdp_tp_integration.py:193]: Is this change needed for `use_orig_params=True` specifically?

File: test/distributed/fsdp/test_fsdp_tp_integration.py [hunk 4]
```diff
@@ -141,31 +141,36 @@ def _sync_tp_grads(
...
-        unsharded_mask = torch.cat(per_param_masks).contiguous().type(torch.BoolTensor)
-        sharded_mask = unsharded_mask.chunk(fsdp_world_size)[self.rank // tp_world_size]
-        grad_device = flat_param.grad.device
-        grad = flat_param.grad.detach().clone().cuda(self.rank)
-        dist.all_reduce(grad, op=dist.ReduceOp.SUM, group=tp_pg)
-        grad = grad.to(grad_device)
-        flat_param.grad[~sharded_mask] = grad[~sharded_mask]
-        # Average *all* gradient elements to match the FSDP only semantics
-        flat_param.grad /= tp_world_size
+        for flat_param in tp_fsdp_model.params:
```
Comment (awgu) [inline test/distributed/fsdp/test_fsdp_tp_integration.py:150]: Is `len(tp_fsdp_model.params) > 1` iff `use_orig_params=True`?

**Review by awgu — APPROVED** [review]
LGTM! We may need to move the `DTensor` import into `_same_storage()` to avoid breaking internal.

File: torch/distributed/fsdp/_flat_param.py (hunk 1 above)
Comment (wz337) [reply torch/distributed/fsdp/_flat_param.py:30]: Is the breakage caused by circular dependency, if you can recall?

Comment (awgu) [reply torch/distributed/fsdp/_flat_param.py:30]: I honestly cannot remember :(

File: test/distributed/fsdp/test_fsdp_tp_integration.py (hunk 2 above)
Comment (weifengpy) [reply test/distributed/fsdp/test_fsdp_tp_integration.py:144]: make it work for `use_orig=True` when `tp_fsdp_model.parameters() > 1`

File: test/distributed/fsdp/test_fsdp_tp_integration.py (hunk 4 above)
Comment (weifengpy) [reply test/distributed/fsdp/test_fsdp_tp_integration.py:150]: that's right

File: test/distributed/fsdp/test_fsdp_tp_integration.py (hunk 3 above)
Comment (weifengpy) [reply test/distributed/fsdp/test_fsdp_tp_integration.py:193]: yes, for `use_orig_params=True` specifically

File: torch/distributed/fsdp/_flat_param.py (hunk 1 above)
Comment (weifengpy) [reply torch/distributed/fsdp/_flat_param.py:30]: got you. will import DTensor inside the function

**Comment by mvpatel2000** [comment]
@weifengpy do you think we can include in torch 2.3.1? 
https://github.com/pytorch/pytorch/issues/125425

**Comment by weifengpy** [comment]
> for FSDP (SHARD_GRAD_OP + use_orig_params) + TP, params in the backward are DTensors. However, `DTensor.untyped_storage().data_ptr()` does not work in `_same_storage`. Thus desugar to `DTensor._local_tensor.untyped_storage().data_ptr()` #123272
> 
> credit to @bigning for the original fix. after landing, we would not need patching in mosaic composer https://github.com/mosaicml/composer/pull/3175/files
> 
> cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang

@mvpatel2000  Just checked I need to cherry-pick this commit otherwiese torch 2.3.1 won't include this fix. Will file a PR to see if we can make it
